### ISSUE TITLE
Constant steps not equal in size

### ISSUE THREAD

**Issue by NinaEffenberger** [probabilistic-numerics/probnum#351]
When solving an ODE with constant steps the very last step might be unnecessary and really small (O(10^-16)) due to rounding errors in the previous steps. E.g. a solver with constant stepsize 0.1 on timespan (0,1) might take 11 steps. @nathanaelbosch @pnkraemer

**Comment by pnkraemer** [comment]
@nathanaelbosch we had an issue about this a while back. I cant find it anymore. What was the solution we agreed on?  Making a step exactly to the endpoint if the new point is larger than `tmax-eps`? 

For what it's worth, here is a fun implementation with the Walrus operator
```python
while (tnew := t + dt) < (tmax - eps):
    # step from t to tnew.
```
unfortunately the Walrus is a 3.8 thing...

**Comment by pnkraemer** [comment]
This might also be affected by #346 .

**Comment by nathanaelbosch** [comment]
We had a discussion on this in #220 with a few ideas, but I don't know what the best thing would be. Ideally I would prefer to hit the endpoint exactly, since you're often interested in evaluating the solution at the endpoint, but I guess that the difference should be negligible if the step is as small as 1e-16.

**Comment by NinaEffenberger** [comment]
The numerical results after the secondlast and last step were the same. To bring up another point: For me it was unsatisfying that the displayed times were not what I set. So the main issue is that too many steps were performed but it would be much nicer if solution.t =[0.1,0.2,0.3,0.4] and not solution.t=[0.1,0.2,0.30000000000001,0.4] (eventhough this is much better than solution.t=[0.1,0.2,0.30000000000001, 0.3999999999999999, 0.4] ). I'm not sure if that solves the problem or brings even more: Is it possible to set the precision of the stepsize to 1e-15? If this is a bad idea, it might be worth a thought to set solution.t =[0.1,0.2,0.3,0.4] afterwards, just talking from the user's point of view.

**Comment by pnkraemer** [comment]
Thanks for the input! 

> The numerical results after the secondlast and last step were the same.

Yes, I think the easiest way would be some sort of corner-case handling, as in "if the distance to the final step is less than 1e-14", move there instead.

> So the main issue is that too many steps were performed but it would be much nicer if solution.t =[0.1,0.2,0.3,0.4] and not solution.t=[0.1,0.2,0.30000000000001,0.4] (eventhough this is much better than solution.t=[0.1,0.2,0.30000000000001, 0.3999999999999999, 0.4] )

Unfortunately, this is due to the way numpy handles floating point numbers and not "our" fault :/ 

> I'm not sure if that solves the problem or brings even more: Is it possible to set the precision of the stepsize to 1e-15

I am not sure whether I get what you mean...

> it might be worth a thought to set solution.t =[0.1,0.2,0.3,0.4] afterwards, just talking from the user's point of view.

This is actually a fun idea (only applicable to constant steps, but there it might indeed be useful. Thoughts @nathanaelbosch ? 

On a side note, this is somewhat related to #346, in the sense that currently, `t_new = t + dt` is set and then inside `step()`, `dt = t_new - t` is inferred. This brings with it some oddities (nothing neckbreaking, but perhaps unintuitive?).

**Comment by nathanaelbosch** [comment]
I think we should not customize too many things here, and in particular not try to change how floats in python work. I agree with the initial point, that getting 11 steps when using a constant 0.1 stepsize on timespan (0,1) is unsatisfying, and we should probably fix that, e.g. with Nico's proposal.

@NinaEffenberger did you have similar issues with scipy, or none at all? Do you have an idea how they handle this topic?

**Comment by NinaEffenberger** [comment]
@pnkraemer I was wondering whether rounding could help (round(stepsize, 15) ). Not a great solution and I'm not sure if that could work, just something that came to my mind...
@nathanaelbosch Scipy uses adaptive steps by default and as far as I know there's no straight-forward way of directly setting the step size in scipy.integrate.

**Comment by pnkraemer** [comment]
True, scipy does not allow fixed steps (unless you're setting `atol` and `rtol` really high and `max_step` fairly low). 

In fact, when artifically forcing fixed steps, scipy seems to have the same problem:
```python
>>> from scipy.integrate import solve_ivp
>>> import numpy as np
>>>
>>> def f(t, x): 
...     return x*(1-x)
>>>
>>> y0 = np.array([0.1])
>>> sol = solve_ivp(f, (0., 1.), y0, atol=1000, rtol=1000, max_step=0.1)
>>> sol.t
array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. , 1. ])
```
